test/cpu_adaptiveavgpool.hpp

```diff
@@ -0,0 +1,311 @@
+    }
+}
+
+template <class T>
+void cpu_adaptiveavgpool_backward_2d(tensor<T> output_grad,
+                                     tensor<T>& input_grad,
+                                     size_t N,
+                                     size_t C,
+                                     size_t H,
+                                     size_t W,
+                                     size_t OH,
+                                     size_t OW)
+{
+    auto dims  = input_grad.desc.GetLengths();
+    auto numel = input_grad.desc.GetElementSize();
+
+    auto output_grad_tv = miopen::get_inner_expanded_tv<4>(output_grad.desc);
+    auto input_grad_tv  = miopen::get_inner_expanded_tv<4>(input_grad.desc);
+
+    for(size_t gid = 0; gid < numel; gid++)
+    {
+        size_t nch = gid / W, w = gid % W;
+        size_t nc = nch / H, h = nch % H;
+        size_t n = nc / C, c = nc % C;
+
+        if(n >= N)
+            return;
+
+        size_t oh  = static_cast<size_t>(std::floor(static_cast<float>(h * OH) / H));
+        size_t koh = static_cast<size_t>(std::ceil(static_cast<float>((h + 1) * OH) / H)) - oh;
+
+        size_t ow  = static_cast<size_t>(std::floor(static_cast<float>(w * OW) / W));
+        size_t kow = static_cast<size_t>(std::ceil(static_cast<float>((w + 1) * OW) / W)) - ow;
+
+        float grad = 0;
+        for(size_t ih = oh; ih < (oh + koh); ++ih)
+        {
+            size_t kh = static_cast<size_t>(std::ceil(static_cast<float>((ih + 1) * H) / OH)) -
+                        static_cast<size_t>(std::floor(static_cast<float>(ih * H) / OH));
+            for(size_t iw = ow; iw < (ow + kow); ++iw)
+            {
+                size_t kw = static_cast<size_t>(std::ceil(static_cast<float>((iw + 1) * W) / OW)) -
+                            static_cast<size_t>(std::floor(static_cast<float>(iw * W) / OW));
+                grad += static_cast<float>(
+                            output_grad[output_grad_tv.get_tensor_view_idx({n, c, ih, iw})]) /
+                        (kh * kw);
+            }
+        }
+
+        input_grad[input_grad_tv.get_tensor_view_idx({n, c, h, w})] = static_cast<T>(grad);
+    }
+}
+
+template <class T>
+void cpu_adaptiveavgpool_backward_3d(tensor<T> output_grad,
+                                     tensor<T>& input_grad,
+                                     size_t N,
+                                     size_t C,
+                                     size_t D,
+                                     size_t H,
+                                     size_t W,
+                                     size_t OD,
+                                     size_t OH,
+                                     size_t OW)
+{
+    auto dims  = input_grad.desc.GetLengths();
+    auto numel = input_grad.desc.GetElementSize();
+
+    auto output_grad_tv = miopen::get_inner_expanded_tv<5>(output_grad.desc);
+    auto input_grad_tv  = miopen::get_inner_expanded_tv<5>(input_grad.desc);
+
+    for(size_t gid = 0; gid < numel; gid++)
```

Comment:
You should use `par_ford` to perform parallel calculation, for better performance.